rename "common" to  "document" & simplify "new window" window logic

deeplow opened the pull request:
Simplify code logic for creating and handling documents through the renaming of `Common` class to `DocumentHolder` and associated simplifications.

When we support bulk document conversion we'll need to keep track of each document's state. This refactor opens way for that. A document holder does not represent a document in itself (e.g. can be instantiated without a document) but stores the input name, output name, and in the future possibly the conversion progress.

Clarification renames: some of the names were changed to better reflect the classes' or files' purpose.
- **renames:** `common.py::Common` -> `document.py::DocumentHandler`
- **renames:** `global_common.py::GlobalCommon` ->`logic.py::DangerzoneCore`
- **renames:** `gui/common.py::GuiCommon` -> `gui/common.py::DangerzoneGui`

Associated logic simplication:
- deduplicates input / output filename validation now done in `DocumentHandler`, when a new filename is added, throwing an `DocumentFilenameException` if that validation fails. (code was previously duplicated in `gui_main` and `cli_main`).
- simplifies "new window" creation logic (and associated `DocumentHandler` instantiation)
   now when the users click *new window*, it simply opens a new window without extra validation to see if any window is already open and doesn't have a document associated to it.

And lastly, it adds unit tests for `document.py` (100% coverage)

deeplow commented:
I was getting errors with the last commit when we ran the cli without a full path. This [is now fixed](https://github.com/freedomofpress/dangerzone/compare/6c615a235f2aed0acb6780287a7d0f547742fc78..c20021b686b60ff5ffa34fddb5d29a90db702499).

```
$ poetry run dev_scripts/dangerzone-cli tests/test_docs/sample.pdf 
╭──────────────────────────╮
│           ▄██▄           │
│          ██████          │
│         ███▀▀▀██         │
│        ███   ████        │
│       ███   ██████       │
│      ███   ▀▀▀▀████      │
│     ███████  ▄██████     │
│    ███████ ▄█████████    │
│   ████████████████████   │
│    ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀    │
│                          │
│    Dangerzone v0.3.2     │
│ https://dangerzone.rocks │
╰──────────────────────────╯

Converting document to safe PDF
> /usr/bin/podman run --network none -u dangerzone --security-opt no-new-privileges --userns keep-id --cap-drop all --rm -v tests/test_docs/sample.pdf:/tmp/input_file -v /home/user/.config/dangerzone/tmp/tmpzkabfaa4/pixels:/dangerzone dangerzone.rocks/dangerzone /usr/bin/python3 /usr/local/bin/dangerzone.py document-to-pixels
Invalid JSON returned from container: Error: error creating named volume "tests/test_docs/sample.pdf": error running volume create option: names must match [a-zA-Z0-9][a-zA-Z0-9_.-]*: invalid argument

documents-to-pixels failed

Failed to convert document
```

deeplow commented on tests/test_document.py:
```diff
@@ -0,0 +1,67 @@
+
+def test_input_file_non_existing():
+    with pytest.raises(DocumentFilenameException):
+        DocumentHolder("fake-dir/non-existing-file.pdf")
```
I need to fix the `/` and make it platform-independent. This won't work on Windows.

deeplow replied:
Done.

apyrgio commented:
I'm testing out this PR in order to see if #204 is resolved, and I have some quick questions:

1. What is the proper way to test this on Windows environments? In 0.3.2, when I click on New Window from the systray, I don't see any window being opened, nor an exception somewhere (I checked the Event Log Viewer as well). I suppose that the exception is logged somewhere, but I haven't found where yet. Any pointers would be helpful.
2. In Linux, I have reproduced the error, and I'm seeing that with your fix, two windows get opened. However, when you close either of them, both of them close, which is unexpected. Am I missing something here?

apyrgio commented:
> I think it would be logged on the terminal from which you launch dangerzone. But I've also had issues where exceptions are consumed somewhere along the codepath. It could be that.

It might be that I'm testing it out by running the `.exe`, not following the instructions on `BUILD.md`. I'll check that out.

> So two windows get open when only one should open? Or that part is working?

I meant that this part is working correctly, yes.

> Yes, that is https://github.com/freedomofpress/dangerzone/issues/205 which was present in linux in 0.3.1. This is fixed in the other PR, I think.

Oh, I thought that #205 was affecting windows that closed due to conversion completion, not due to a user action. Ok, then we'll see this later on.

apyrgio commented on dangerzone/gui/__init__.py:
```diff
@@ -85,50 +86,42 @@ def delete_window(window_id: str) -> None:
-            len(windows) == 1
-            and windows[list(windows.keys())[0]].document.input_filename == None
-        ):
-            window = windows[list(windows.keys())[0]]
```
If I understand correctly, we now remove this logic whereby only one window could get created. This should be documented in the commit message that introduces this removal, since currently we are just saying we are simplifying the new window logic.

deeplow replied:
I don't think so. I think here the goal was really just to simplify the logic. It can be quite convoluted to understand but from my understanding, before it was checking if there was already a window open without a selected document. If there was, it would not create a window. If there wasn't then it would create a new window.

This check didn't seem important as what is the consequence of the user starting multiple windows without.

apyrgio replied:
I see. I didn't realize there could be a case where you have a window open without a document. If it's not important then sure, it can go.

apyrgio commented on dangerzone/document.py:
```diff
@@ -12,9 +12,9 @@ class Document(object):
     """
 
-    def __init__(self) -> None:
+    def __init__(self, input_file_path: str) -> None:
```
The "simplify 'new window' logic" commit tries to do three things at once:

1. Change the way we instantiate the document class.
2. Moves the filename checks out of the new window function.
3. Simplifies the window logic a bit.

Specifically for (2), we move this logic again under the `DocumentHolder()` class in a subsequent commit. So, all in all, it's better to break these three things in different commits, and merge them with any amendments we do in subsequent commits. This way, each commit has a clear purpose, and it doesn't introduce logic that will be altered.

apyrgio replied:
Great, thanks for the explanation.

apyrgio commented on dangerzone/cli.py:
```diff
@@ -31,51 +31,23 @@ def cli_main(
-            exit(1)
+    except DocumentFilenameException as e:
+        click.echo(str(e))
+        exit(1)
```
Why do we need to catch `DocumentFilenameException` in various places in the code? Can we catch them in the main function instead?

deeplow replied:
This is done in the child PR #216. [These lines](https://github.com/freedomofpress/dangerzone/pull/216/commits/3be0b852efeaf784eae826a0e17aa92ce6e1c1ad#diff-1818f923cb4b1458022430fcaea85123d0a5d8fd15b270705b5ba7ba085938deL33-L46) specifically.

I'd prefer avoiding further complicating that rebase.

apyrgio replied:
Got it, thanks.

apyrgio commented on dangerzone/document.py:
```diff
@@ -12,29 +12,57 @@ class DocumentHolder(object):
+                pass
+        except PermissionError:
+            raise DocumentFilenameException("Safe PDF filename is not writable")
+
```
I fully agree with moving these checks closer to the `DocumentHolder` class. However, this way we can't reuse them from other contexts. For instance, it would make the CLI logic much simpler if we used Click validators: https://click.palletsprojects.com/en/8.1.x/options/#callbacks-for-validation. If we moved this logic into class/static methods, it would be trivial to create those.

deeplow replied:
I tried making this a static method, but it will look weird because click's validators have three parameters (`ctx`, `param`, `value`) and not just `filename`. And we'd have to duplicate this validation code because at this moment we click for the cli and another for the gui.

apyrgio replied:
I believe this could be solved via:

* An `args.py` module that is Click-aware,
* Some `validate_{input,output}_filename()` functions, that implement the callback interface for Click (`ctx`, `param`, `value`) and wrap the `Document.validate_{input,output}_filename()` methods.
* Calling the `args.validate_{input,output}_filename()` functions from the CLI and GUI code.

deeplow replied:
Sounds good. Yes, if you could write it, that would work for me.

apyrgio commented on dangerzone/document.py:
```diff
@@ -12,29 +12,57 @@ class DocumentHolder(object):
+
+class DocumentFilenameException(Exception):
+    def __init__(self, message):
+        super().__init__(message)
```
Also, the `DocumentFilenameException()` class is introduced in 'simplify "new window" logic", and is patched in "add unit tests for document.py". It's best to keep thing clean and merge the patch in the commit that introduces this class.

deeplow replied:
> Also, the DocumentFilenameException() class is introduced in 'simplify "new window" logic", and is patched in "add unit tests for document.py". 

Now it's introduced in `validate I/O filenames in Document class` and is no longer patched there.

> I'm not sure about the `DocumentFilenameException` as is,

This a UX thing. Front ends (closer to the user) should present more actionable information. Otherwise, on every front end we'll have to catch the original exception and show the user-understandable translation. But using an intermediary exception
  * Instead of showing the error: `[Errno 2] No such file or directory: document.pdf` which is not user-actionable
  * instead we show: `Input file not found: make sure you typed it correctly.`


> because we lose traceback info and the original message. We could retain these with raise ... from ..., but is it worth the extra complexity?

I have added the `raise ... from ...` syntax. I was not aware this was a thing.

apyrgio replied:
Cool, thanks for the changes. As for the original traceback, now that we retain it with `raise ... from ...`, it would be nice logging it as well. This requires handling the `DocumentFilenameException` from a single place. I'll give it a try.

apyrgio commented on tests/test_document.py:
```diff
@@ -0,0 +1,67 @@
+    Attempts to read a document's filename when no doc has been set
+    """
+    d = DocumentHolder()
+    with pytest.raises(DocumentFilenameException):
```
When testing for exceptions, it's good to validate the exception message as well, especially for classes like `DocumentFilenameException` that can have multiple messages, to verify that the expected check was triggered. See `match=` in https://docs.pytest.org/en/7.1.x/how-to/assert.html#assertions-about-expected-exceptions13.

apyrgio commented on dangerzone/logic.py:
```diff
@@ -18,9 +18,9 @@
+class DangerzoneCore(object):
     """
-    The GlobalCommon class is a singleton of shared functionality throughout the app
+    Singleton of shared state / functionality throughout the app
```
`Dangerzone{Core,Gui}` is basically a collection of common functions and state for:

* Graphics (terminal colors, fonts, window icon)
* Settings (app path, OCR languages)
* Qt app
* Docker status

The fact that it's used for so many different stuff makes it candidate for becoming bloated, and not easily shareable between contexts. I understand that this class (and the common ones before it) act as an easy way to share the above logic between different contexts, but I believe we would greatly benefit if we moved the above logic into their own modules. We would be able to unit test them, and use them from any context, without having to pass around the Dangerzone class.

Note that we definitely need a "context-like" class, which is used to store the state for the conversion operation as a whole, but that's a separate thing. Also, in this PR I don't see this class holding any of that state, which is why I'm still not sure why it's needed. I believe it's used in subsequent PRs, so in this case, it would be more clear if we explain why we want it in the commit message.

apyrgio replied:
Understandable. I'm bringing it up now so that we don't overload this class in subsequent PRs. But we can see that later on.

apyrgio commented:
Also, I see the following commits:

```
Common -> Document                                                                             
Document -> DocumentHolder                                                                     
rename (Global)Common & related to "dangerzone"                                                
rename global_common.py -> logic.py
```

Might if I make their names a bit more consistent, such as "Rename X module/class to Y"?

deeplow commented on dangerzone/document.py:
```diff
@@ -7,9 +7,9 @@
 
 
-class Document(object):
+class DocumentHolder(object):
```
1. is now done (the `Document` -> `DocumentHolder`) commit was deleted (in a parallel branch)
2. this is is better left for the multi-document PR. I tried that as well here, but there are situations in which the Document class isn't set in advance with the document name.

deeplow commented:
@apyrgio I have addressed all comments in the parallel branch [simplify-common-2](https://github.com/freedomofpress/dangerzone/tree/simplify-common-2)

apyrgio commented:
@deeplow Thanks a lot for the changes. I'll take a look and report back.

apyrgio commented:
I have done some changes of my own, based on the above discussion, which you can find on the `simplify-common-3` branch. I have tested it on Linux and it works. I need to iron out one small detail, and then it will be good to merge.

deeplow commented:
Whoops. I have accidentally deleted this while trying to force push.

deeplow commented:
- Rebased w/ newly included windows tests
- added type hints to `test_documents.py` (because tests now have type hinting requirements)
- fixed `args.py` and `errors.py` type hints

@apyrgio I think the last thing missing is you changing your author email, if you so desire.

deeplow commented:
Interesting. Mypy is throwing some errors but it doesn't throw them on my system.

````
dangerzone/errors.py:4: error: Module "typing" has no attribute "ParamSpec"; maybe "_ParamSpec"?  [attr-defined]
dangerzone/errors.py:18: error: The first argument to Callable must be a list of types, parameter specification, or "..."  [misc]
dangerzone/errors.py:18: note: See https://mypy.readthedocs.io/en/stable/kinds_of_types.html#callable-types-and-lambdas
dangerzone/errors.py:22: error: Name "P.args" is not defined  [name-defined]
dangerzone/errors.py:22: error: Name "P.kwargs" is not defined  [name-defined]
```

If we take a look at the docs, it is well documented there: https://mypy.readthedocs.io/en/stable/generics.html?highlight=ParamSpec#declaring-decorators

deeplow commented:
It looks like this happening because some of CI doesn't have python3.10 version running. We might have moved too soon with https://github.com/freedomofpress/dangerzone/pull/161.

deeplow commented:
Rebased and ready to be merged. Can I merge it @apyrgio?

apyrgio commented:
Gimme a bit to take a last look, and running the unit tests on Windows. Also, we need to squash "FIXUP" commits, since they pollute the history.

deeplow commented:
> Also, we need to squash "FIXUP" commits, since they pollute the history.

Done

apyrgio commented:
I'm doing some final changes (proper author attribution to some of the commits), and I'm testing on a Windows machine. I'll ping you when you can rebase, sign, and merge.

deeplow commented on tests/test_document.py:
```diff
@@ -0,0 +1,80 @@
+# See also:
+# https://stackoverflow.com/questions/72528318/what-file-permissions-make-a-file-unreadable-by-owner-in-windows
+@pytest.mark.skipif(platform.system() == "Windows", reason="Unix-specific")
+def test_input_file_unreadable(unreadable_pdf):
```
@apyrgio you're only missing here the type hints.

apyrgio replied:
Yeap, I'm on it as we speak.

apyrgio replied:
Done and force pushed.

apyrgio commented:
I've updated the branch with the following changes:

1. I properly attributed some commits to you. I had split some commits while reviewing and the author became myself, but I changed that back.
2. I fixed a Windows unit test error that I had actually fixed before on `simplify-common-4`. We probably didn't get the latest changes before force-pushing to `simplify-common`. That's ok, but in general it's good to have this in mind.
3. I checked that the Windows unit tests do pass. The only (minor) issue that we have is that `test_lang_eng` fails, because Tesseract times out. Since this test works on Linux, and Tesseract runs on a container, I'll attribute it to our timeouts configuration and my slow Windows VM.
